client/src/ui/BitbakeCommands.ts

```diff
@@ -16,22 +16,30 @@ import { extractRecipeName } from '../lib/src/utils/files'
 import { runBitbakeTerminal, runBitbakeTerminalCustomCommand } from './BitbakeTerminal'
 import { type BitbakeDriver } from '../driver/BitbakeDriver'
 import { sanitizeForShell } from '../lib/src/BitbakeSettings'
-import { type BitbakeTaskDefinition, type BitbakeTaskProvider } from './BitbakeTaskProvider'
+import { type BitbakeCustomExecution, type BitbakeTaskDefinition, type BitbakeTaskProvider } from './BitbakeTaskProvider'
 import { type LayerInfo } from '../lib/src/types/BitbakeScanResult'
 import { DevtoolWorkspaceTreeItem } from './DevtoolWorkspacesView'
 import { finishProcessExecution } from '../lib/src/utils/ProcessUtils'
 import { type SpawnSyncReturns } from 'child_process'
 import { clientNotificationManager } from './ClientNotificationManager'
 import { configureDevtoolSDKFallback } from '../driver/BitbakeESDK'
+import { type LanguageClient } from 'vscode-languageclient/node'
+import { RequestMethod } from '../lib/src/types/requests'
 
 let parsingPending = false
 let bitbakeSanity = false
 
+let _languageClient: LanguageClient
```

Comment:
not a fan of singletons in the client which has a lot less of them. I suggest to rather mirror the BitbakeProjectScanner for the client part.

1. Create a BitbakeRecipeScanner class
2. It should have a function scanRecipe() which starts the task, that's what should be called by the BitbakeCommand.ts
3. Have it's own onDidEndTask() handler 
4. The handler calls a parseBitbakeEnvironment() which takes the `-e` log and generates a new type: BitbakeRecipeScanResult (mirroring BitbakeScanResult)
5.  BitbakeRecipeScanner has a method setLanguageClient() to prepare sending the notification (not request!) (type BitbakeScanResult, not string)
6. You could move the instance of BitbakeRecipeScanner as an attribute of bitbakeProjectScanner for easier manipulation in extension.ts

That way we have a uniformed scanning logic for recipes that is the same as the project scan, and a much more focused class for handling recipe scans.

Reply:
This commit contains the changes you mentioned: 382f61cfaa300001818838777d473c3352066567
But If we do the parsing in the server then we might still use the request so it can return the data to the client for further usage.

Reply:
For now, I will parse the change history for the variables that exist in the current file. Let me know what else to parse.

Reply:
I guess the main concern is you want to replace the parseAllRecipe command with this scan command, which will make sending the results a heavy task since they are usually huge. How about we keep both of them? Let `parseAllRecipes` do what it has been doing and let the scan command be an alternative.

Reply:
If the server has the tools you need to parse then it's a valid reason to do the parsing in the server. The other alternative would be to move it in the common library.

Reply:
Try implementing a full parsing and see if it really hits performance on save? If it does, we can do a lazy parsing instead: Do a minimal first parsing, then per variable parsing on request?
Tree sitter parsing is supposed to be fast enough to be done on typing event! I don't foresee it to be an issue on save. Replacing parseAllRecipes on save with the recipe environment scan will be really huge in terms of functionality.

Reply:
You can find a scan features list in our internal tracking ticket:

    Provide hover for variables with their final value
    Provide definition with the accurate list of assignments for a variable
    Provide more accurate definitions for require/inherit statements (If result is not already 100% correct?)
    Provide accurate recipe includes in the bitbake recipe view (left pannel). Could become bloated so select the most important ones?
    Provide definition for override operators (go to the corresponding .conf from the include list given by bitbake -e)

Reply:
1. I will work on the parsing on the server side first then see if it should be moved to the common lib
2. Yes, Tree-sitter is fast and it gives results right after every type. 
3. Let's go for the lazy parse approach as you said for now.

And I will move the implementation for applying the scan result to another PR(s). Let this one be the preparation stage